mobile_config/src/client/gateway_client.rs

```diff
@@ -0,0 +1,110 @@
+use helium_proto::{
+    services::{mobile_config, Channel},
+    Message,
+};
+use retainer::Cache;
+use std::{sync::Arc, time::Duration};
+
+#[derive(Clone)]
+pub struct GatewayClient {
+    pub client: mobile_config::GatewayClient<Channel>,
+    signing_key: Arc<Keypair>,
+    config_pubkey: PublicKey,
+    batch_size: u32,
+    cache: Arc<Cache<PublicKeyBinary, Option<gateway_info::GatewayInfo>>>,
+    cache_ttl: Duration,
+}
+
+impl GatewayClient {
+    pub fn from_settings(settings: &Settings) -> Result<Self, Box<helium_crypto::Error>> {
+        let cache = Arc::new(Cache::new());
+        let cloned_cache = cache.clone();
+        tokio::spawn(async move {
+            cloned_cache
+                .monitor(4, 0.25, CACHE_EVICTION_FREQUENCY)
+                .await
+        });
+
+        Ok(Self {
+            client: settings.connect_gateway_client(),
+            signing_key: settings.signing_keypair()?,
+            config_pubkey: settings.config_pubkey()?,
+            batch_size: settings.batch_size,
+            cache_ttl: settings.cache_ttl(),
+            cache,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl gateway_info::GatewayInfoResolver for GatewayClient {
+    type Error = ClientError;
+
+    async fn resolve_gateway_info(
+        &mut self,
+        address: &PublicKeyBinary,
+    ) -> Result<Option<gateway_info::GatewayInfo>, Self::Error> {
+        if let Some(cached_response) = self.cache.get(address).await {
+            return Ok(cached_response.value().clone());
+        }
+
+        let mut request = mobile_config::GatewayInfoReqV1 {
+            address: address.clone().into(),
+            signer: self.signing_key.public_key().into(),
+            signature: vec![],
+        };
+        request.signature = self.signing_key.sign(&request.encode_to_vec())?;
+        tracing::debug!(pubkey = address.to_string(), "fetching gateway info");
+        let response = match self.client.info(request).await {
+            Ok(info_res) => {
+                let response = info_res.into_inner();
+                response.verify(&self.config_pubkey)?;
+                response.info.map(gateway_info::GatewayInfo::from)
+            }
+            Err(status) if status.code() == tonic::Code::NotFound => None,
+            Err(status) => Err(status)?,
+        };
+
+        self.cache
+            .insert(address.clone(), response.clone(), self.cache_ttl)
+            .await;
+
+        Ok(response)
+    }
+
+    async fn stream_gateways_info(
+        &mut self,
+    ) -> Result<gateway_info::GatewayInfoStream, Self::Error> {
+        let mut req = mobile_config::GatewayInfoStreamReqV1 {
+            batch_size: self.batch_size,
+            signer: self.signing_key.public_key().into(),
+            signature: vec![],
+        };
+        req.signature = self.signing_key.sign(&req.encode_to_vec())?;
+        tracing::debug!("fetching gateway info stream");
+        let pubkey = Arc::new(self.config_pubkey.clone());
```

Comment:
the Arc'ing here seems redundant

Reply:
In this case we need to b/c we’re moving the cloned key into the async closure within the stream to verify the signature on each batch of gateways returned. We have to clone it the once to copy it out of the client itself but then we move the it arc-wrapped into the closure

Reply:
I tried it locally and clippy doesnt highlight any probs in the closure with the arc removed